File: librarymanager/librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/BookUsingEntity.java
package ru.elenacherev.librarymanager.domain.entity;

import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;

import static javax.persistence.GenerationType.IDENTITY;

@Entity
@Table(name = "BOOK_USINGS")
@Getter
@NoArgsConstructor
//@RequiredArgsConstructor
@Setter
public class BookUsingEntity {
    @Id
    @GeneratedValue(strategy = IDENTITY) //уникальный ID, генерируется сам в БД
    @Column(name = "BOOK_USING_ID")
    @NotNull
    @Setter(AccessLevel.NONE)
    private Long bookUsingId;

    @Version
    @Column(name = "VERSION")
    private int version; //версия, инкрементится при редактировании записи

    @NotNull(message = "{validation.startDate.NotEmpty.message}")
    @Temporal(TemporalType.DATE)
    @Column(name = "START_DATE")
    private Date startDate; //Книгу взяли

    @Temporal(TemporalType.DATE)
    @Column(name = "END_DATE")
    private Date endDate; //Книгу вернули

    @ManyToOne
    @JoinColumn(name = "READER_ID")
    private ReaderEntity reader; //кто взял книгу

    @ManyToOne
    @JoinColumn(name = "BOOK_ID")
    private BookEntity book; //какую книгу взяли

}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/domain/entity/BookUsingEntity.kt
package ru.elenacherev.librarymanager.domain.entity

import org.hibernate.annotations.ColumnDefault
import java.time.Instant
import java.util.*
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.Table
import javax.persistence.Temporal
import javax.persistence.TemporalType
import javax.persistence.Version
import javax.validation.constraints.NotNull

@Entity
@Table(name = "BOOK_USINGS")
data class BookUsingEntity(

    @Column(name = "BOOK_USING_ID")
    @GeneratedValue
    @Id
    @get:NotNull
    var bookUsingId: UUID? = null,

    //версия, инкрементится при редактировании записи
    @Version
    @Column(name = "VERSION")
    var version: Int = 0,

    //Книгу взяли
    @Column(name = "START_DATE")
    @NotNull
    var startDate: Instant,

    //Книгу вернули
    @Column(name = "END_DATE")
    var endDate: Instant? = null,

    //кто взял книгу
    @ManyToOne
    @JoinColumn(name = "READER_ID")
    var reader: ReaderEntity,

    //какую книгу взяли
    @ManyToOne
    @JoinColumn(name = "BOOK_ID")
    var book: BookEntity
)